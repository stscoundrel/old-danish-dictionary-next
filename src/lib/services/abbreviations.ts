import { findAbbreviations } from 'old-danish-dictionary'
import { abbreviate } from 'abbreviatrix'
import { OriginalDictionaryEntry } from 'lib/models/dictionary'

export interface Abbreviation{
    abbreviation: string,
    explanation: string
}

export const getAbbreviations = ({ definitions }: OriginalDictionaryEntry): Abbreviation[] => {
  const combinedAbbreviations: Abbreviation[] = []
  const abbreviationSet = new Set()

  definitions.forEach((definition) => {
    const abbreviations = findAbbreviations(definition)
    abbreviations.forEach((explanation, abbreviation) => {
      if (!abbreviationSet.has(abbreviation)) {
        abbreviationSet.add(abbreviation)
        combinedAbbreviations.push({ abbreviation, explanation })
      }
    })
  })

  return combinedAbbreviations;
}

/**
 * Add abbr tags to content with explanations.
 */
export const addAbbreviationsToContent = (
  content: string,
  abbreviations: Abbreviation[],
): string => {
  let result = content

  abbreviations.forEach(({ abbreviation, explanation }) => {
    result = abbreviate(abbreviation, explanation, result)
  })

  return result
}
