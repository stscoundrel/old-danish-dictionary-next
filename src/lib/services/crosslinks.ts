import { getCrosslinks as getAllCrossLinks, Crosslink, getOldDanishCrosslinks } from 'scandinavian-dictionary-crosslinker'
import { DictionaryEntry } from 'lib/models/dictionary'

export type { Crosslink } from 'scandinavian-dictionary-crosslinker'

let crossLinkCache: Record<string, Crosslink[]> | null = null

const populateCrossLinks = () => {
  if (!crossLinkCache) {
    crossLinkCache = getAllCrossLinks()
  }

  return crossLinkCache
}

export const getCrossLinks = (entry: DictionaryEntry): Crosslink[] => {
  const crosslinks = populateCrossLinks()
  if (Object.prototype.hasOwnProperty.call(crosslinks, entry.slug)) {
    return getOldDanishCrosslinks(entry.slug)
  }

  return []
}
