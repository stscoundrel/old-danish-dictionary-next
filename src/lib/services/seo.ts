import { joinWithConj } from 'teljari'
import { capitalize } from 'lib/utils/strings'
import { DictionaryEntry, DictionaryEntryDTO } from 'lib/models/dictionary'
import { AlphabetLetter } from 'lib/services/dictionary'

export interface SEO {
  title: string,
  description: string
}

export const getLetterSEO = (
  alphabetLetter: AlphabetLetter,
  content: DictionaryEntry[] | DictionaryEntryDTO[],
): SEO => {
  const firstWords = content.slice(0, 4).map((entry) => entry.headword.toLowerCase())
  return {
    title: `Old Danish words starting with letter ${alphabetLetter.letter.toUpperCase()}`,
    description: `Meanings of Old Danish words starting with "${firstWords[0].charAt(0).toUpperCase()}", such as ${joinWithConj(firstWords)}`,
  }
}

export const getWordSEO = (word: DictionaryEntry): SEO => ({
  title: `Old Danish Dictionary - ${capitalize(word.headword)}`,
  description: `Meaning of Old Danish word "${word.headword.toLowerCase()}" in Danish`,
})

export const getDefaultSEO = () : SEO => ({
  title: 'Old Danish Dictionary - Otto Kalkar',
  description: '45 000+ words and definitions for old, middle and early modern danish.',
})
