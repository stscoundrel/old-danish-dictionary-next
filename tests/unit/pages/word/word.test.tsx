import ReactDOM from 'react-dom/client'
import Word, { getStaticProps, getStaticPaths } from 'pages/word/[word]'
import renderer from 'react-test-renderer'
import { getAlphabet } from 'lib/services/dictionary'
import { DictionaryEntry } from 'lib/models/dictionary'

const mockHandler = jest.fn()

/**
 * Mock router
 */
jest.mock('next/router', () => ({
  useRouter() {
    return {
      locale: undefined,
      defaultLocale: undefined,
      asPath: '/test',
      back: mockHandler,
    }
  },
}))

const entry: DictionaryEntry = {
  headword: 'Abe',
  definitions: [
    'no.',
    "1) dyrenavn, hunk. (cv sv. apa.); som aben tit skeer, naar hun vil gøre effter det, hun seer. Herm. ' Weigere. 215, 228%; lige som man ick kunde aben kende, naar hun sig før i Aarons klæder. Ranch. 29; løb løss en tam abe, der mand lockede ad hende. Hvitf VIL 1; aben, som bilder sig ind, at hendis børn er altijd de smuckeste. S. Terkelsen, Astreæ Sjungek. IL. fort. —",
    '2) fjfr. isl. api.) dåre; wthen han seer ther tijl ful- breth, tha holles han forenabe., Rimkr. p%; giør du Joab, Herre, til din abe, at hans anslag ey due. Ranch. 95; en nar, den, som holder for- meget af sine børn. Moth.',
  ],
  slug: 'abe',
}

const abbreviations = [
  { abbreviation: 'no.', explanation: 'navneord (substantivum).' },
  { abbreviation: 'hunk.', explanation: 'hunkøn (femininum).' },
  { abbreviation: 'sv.', explanation: 'svensk.' },
  { abbreviation: 't.', explanation: 'tysk.' },
  { abbreviation: 'isl.', explanation: 'islandsk.' },
  { abbreviation: 'n.', explanation: 'norsk.' },
]

const letter = {
  letter: 'a',
  slug: 'a',
}

describe('Word page: render & usage', () => {
  test('Does not crash', () => {
    const div = document.createElement('div')
    const root = ReactDOM.createRoot(div)
    root.render(
      <Word
        entry={entry}
        similarEntries={[]}
        letters={getAlphabet()}
        letter={letter}
        abbreviations={abbreviations}
      />,
    )
  })

  test('Matches snapshot', () => {
    const tree = renderer.create(
      <Word
        entry={entry}
        similarEntries={[]}
        letters={getAlphabet()}
        letter={letter}
        abbreviations={abbreviations}
      />,
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('Returns null if entry is unavailable', () => {
    const tree = renderer.create(
      <Word
        entry={null}
        similarEntries={[]}
        letters={getAlphabet()}
        letter={letter}
        abbreviations={abbreviations}
      />,
    ).toJSON()
    expect(tree).toBeNull()
  })

  test('Back button works', async () => {
    const tree = renderer.create(
      <Word
      entry={entry}
      similarEntries={[]}
      letters={getAlphabet()}
      letter={letter}
      abbreviations={abbreviations}
    />,
    )

    // Click back btn.
    await renderer.act(async () => {
      expect(mockHandler).not.toHaveBeenCalled()
      await tree.root.findByProps({ text: 'Back' }).props.action()

      // Assert mockrouter received a push.
      expect(mockHandler).toHaveBeenCalled()
      expect(mockHandler.mock.calls.length).toBe(1);
    })
  })
})

describe('Word page: data fetching', () => {
  test('getStaticPaths works', async () => {
    const expected = {
      paths: [],
      fallback: 'blocking',
    }

    const result = await getStaticPaths()

    expect(result).toMatchObject(expected)
  })

  test('getStaticProps works', async () => {
    const expected = {
      props: {
        entry,
        similarEntries: [
          {
            headword: 'Abe',
            definitions: [
              'go.',
              "1) skabe sig (efter); somdet papistiske folok haffuer nock abet sig. Pallad., S. Ped. Skib. r4'; nu grøn, nu blaa, mu sort for herre- gunst sig aber. Hexaem. 243; abe sig efter alle dyr med munden. Grev. 0. Frih. K. 50. —",
              '2) narre, spotte; naar hand haffuer dig behoft, da kand han smucht abe dig. Jes. Sir. 13.7 (Luth. åffen); huad haffue i voss her atabe. Herm, Weigere. 255; paa det sidste aff lycken abit oc bedaarit. Hvitf. VIL 1; har han stillet sig,som hand nødig vilde komme oc abet dennem med dette giensvar. Bruns- mand, Huus-Kaars. 1757. 74; smlen. Sch. u. L.: apen og affen.',
            ],
            slug: 'abe-2',
          },
        ],
        letters: getAlphabet(),
        letter,
        abbreviations,
      },
    }

    const result = await getStaticProps({ params: { word: 'abe' } })

    expect(result).toEqual(expected)
  })

  test('getStaticProps returns 404 redirect for unkown words', async () => {
    const expected = {
      props: {},
      notFound: true,
    }

    const result = await getStaticProps({ params: { word: 'loremipsum' } })

    expect(result).toEqual(expected)
  })
})
