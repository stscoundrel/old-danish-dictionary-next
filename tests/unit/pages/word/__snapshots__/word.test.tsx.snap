// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Word page: render & usage Matches snapshot 1`] = `
[
  <header>
    <div
      className="topBarSpacer"
    />
    <nav
      className="section"
    >
      <div
        className="topbar container"
      >
        <a
          href="/"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onTouchStart={[Function]}
        >
          <img
            alt="To home"
            data-nimg="1"
            decoding="async"
            height={30}
            loading="lazy"
            onError={[Function]}
            onLoad={[Function]}
            src="/_next/image?url=%2Ffavicon-48x48.png&w=64&q=75"
            srcSet="/_next/image?url=%2Ffavicon-48x48.png&w=32&q=75 1x, /_next/image?url=%2Ffavicon-48x48.png&w=64&q=75 2x"
            style={
              {
                "color": "transparent",
              }
            }
            width={30}
          />
        </a>
        <div
          aria-label="Open menu"
          className="hamburger"
          onClick={[Function]}
          role="button"
        >
          <span
            className="part"
          />
          <span
            className="part"
          />
          <span
            className="part"
          />
        </div>
      </div>
      <div
        className="content null container"
      >
        <ul
          className="list"
        >
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/a"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              a
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/b"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              b
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/d"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              d
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/e"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              e
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/f"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              f
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/g"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              g
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/h"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              h
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/i"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              i
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/j"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              j
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/k"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              k
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/l"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              l
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/m"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              m
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/n"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              n
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/o"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              o
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/p"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              p
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/r"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              r
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/s"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              s
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/t"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              t
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/u"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              u
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/v"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              v
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/x"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              x
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/y"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              y
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/ae"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              æ
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/oe"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              ø
            </a>
          </li>
        </ul>
      </div>
    </nav>
    <nav
      className="section"
    >
      <div
        className="container"
      >
        <a
          className="link"
          href="/"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onTouchStart={[Function]}
        >
          Old Danish Dictionary
        </a>
        <a
          className="link"
          href="/letter/a"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onTouchStart={[Function]}
        >
          Letter A
        </a>
        <a
          className="link"
          href="/word/abe"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onTouchStart={[Function]}
        >
          Abe
        </a>
        <script
          dangerouslySetInnerHTML={
            {
              "__html": "{"@context":"https://schema.org/","@type":"BreadcrumbList","itemListElement":[{"@type":"ListItem","position":1,"name":"Old Danish Dictionary","item":"undefined/"},{"@type":"ListItem","position":2,"name":"Letter A","item":"undefined/letter/a"},{"@type":"ListItem","position":3,"name":"Abe","item":"undefined/word/abe"}]}",
            }
          }
          type="application/ld+json"
        />
      </div>
    </nav>
  </header>,
  <main
    className="container"
  >
    <article
      className="section"
    >
      <header>
        <h1
          lang="non"
        >
          Abe
        </h1>
        <small
          className="subHeading"
        >
          Old Danish Dictionary - 
          abe
        </small>
        <p>
          Meaning of Old Danish word 
          <em>
            "
            Abe
            "
          </em>
          , as defined by Otto Kalkar's Dictionary of Old Danish language.
        </p>
      </header>
      <p>
        The Old Danish word 
        <strong>
          <dfn
            className="capitalize"
          >
            Abe
          </dfn>
        </strong>
         can mean:
      </p>
      <ul
        className="definitionList"
      >
        <li
          className="definitionItem"
          dangerouslySetInnerHTML={
            {
              "__html": "<abbr title="navneord (substantivum).">no.</abbr>",
            }
          }
          lang="swe"
        />
        <li
          className="definitionItem"
          dangerouslySetInnerHTML={
            {
              "__html": "1) dyrenavn, <abbr title="hunkøn (femininum).">hunk.</abbr> (cv <abbr title="svensk.">sv.</abbr> apa.); som aben tit skeer, naar hun vil gøre effter det, hun seer. Herm. ' Weigere. 215, 228%; lige som man ick kunde aben kende, naar hun sig før i Aarons klæder. Ranch. 29; løb løss en tam abe, der mand lockede ad hende. Hvitf VIL 1; aben, som bilder sig ind, at hendis børn er altijd de smuckeste. S. Terkelsen, Astreæ Sjungek. IL. for<abbr title="tysk.">t.</abbr> —",
            }
          }
          lang="swe"
        />
        <li
          className="definitionItem"
          dangerouslySetInnerHTML={
            {
              "__html": "2) fjfr. <abbr title="islandsk.">isl.</abbr> api.) dåre; wthen han seer ther tijl ful- breth, tha holles han forenabe., Rimkr. p%; giør du Joab, Herre, til din abe, at hans anslag ey due. Ranch. 95; en nar, den, som holder for- meget af sine bør<abbr title="norsk.">n.</abbr> Moth.",
            }
          }
          lang="swe"
        />
      </ul>
      <p>
        Possible runic inscription in 
        <em>
          Medieval Futhork
        </em>
        :
        <span
          className="rune"
        >
          ᛆᛒᚽ
        </span>
        <br />
        <small>
          Medieval Runes were used in Denmark from 12th to 17th centuries.
        </small>
      </p>
      <div
        className="abbreviations"
      >
        <div
          className="column"
        >
          <h4>
            Abbreviations used:
          </h4>
          <dl
            className="wrap"
          >
            <dt
              className="abbreviation"
            >
              <strong>
                no.
              </strong>
            </dt>
            <dd
              className="explanation"
            >
              navneord (substantivum).
            </dd>
          </dl>
          <dl
            className="wrap"
          >
            <dt
              className="abbreviation"
            >
              <strong>
                hunk.
              </strong>
            </dt>
            <dd
              className="explanation"
            >
              hunkøn (femininum).
            </dd>
          </dl>
          <dl
            className="wrap"
          >
            <dt
              className="abbreviation"
            >
              <strong>
                sv.
              </strong>
            </dt>
            <dd
              className="explanation"
            >
              svensk.
            </dd>
          </dl>
          <dl
            className="wrap"
          >
            <dt
              className="abbreviation"
            >
              <strong>
                t.
              </strong>
            </dt>
            <dd
              className="explanation"
            >
              tysk.
            </dd>
          </dl>
          <dl
            className="wrap"
          >
            <dt
              className="abbreviation"
            >
              <strong>
                isl.
              </strong>
            </dt>
            <dd
              className="explanation"
            >
              islandsk.
            </dd>
          </dl>
          <dl
            className="wrap"
          >
            <dt
              className="abbreviation"
            >
              <strong>
                n.
              </strong>
            </dt>
            <dd
              className="explanation"
            >
              norsk.
            </dd>
          </dl>
        </div>
      </div>
    </article>
    <a
      className="button"
      onClick={[Function]}
    >
      Back
    </a>
    <div
      aria-label="Back to top"
      className="backToTop"
      onClick={[Function]}
      role="button"
    >
      ↑
    </div>
  </main>,
  <footer
    className="section"
  >
    <div
      className="container"
    >
      <section
        className="section"
      >
        <h2>
          About
        </h2>
        <p>
          Based on "Ordbog til det ældre danske Sprog" by Otto Kalkar.
        </p>
        <p>
          It is the largest Old Danish dictionary. The original volumes were published in 1881 - 1907, with additional supplement added in 1907 - 1918.
        </p>
      </section>
      <section
        className="section"
      >
        <h3
          className="h4"
        >
          Old Danish language
        </h3>
        <p>
          Old Swedish developed from Old East Norse, the eastern dialect of Old Norse, at the end of the Viking Age.
        </p>
        <p>
          Despite its name, Kalkars dictionary is not exactly a dictionary of "Old Danish",as it covers period from Middle Danish to early Modern Danish. The dictionary covers time period of 1300 - 1700, which means the oldest vocabulary would not differ much from that used in the Viking Age, while the newest portions would not differ much from modern Danish.
        </p>
      </section>
      <section
        className="section"
      >
        <h4>
          Old Norse language
        </h4>
        <p>
          Old Norse was a North Germanic language that was spoken by inhabitants of Scandinavia and their overseas settlements from about the 7th to the 15th centuries.
        </p>
        <p>
          Also known as "the viking language", "Old Nordic", or "Old Scandinavian"
        </p>
      </section>
      <div
        className="navs"
      >
        <nav
          className="nav"
        >
          <h4
            className="navTitle"
          >
            Dictionary project
          </h4>
          <ul>
            <li>
              <a
                href="https://github.com/stscoundrel/old-danish-dictionary-next"
                rel="noopener noreferrer"
                target="_blank"
              >
                Source code
              </a>
            </li>
            <li>
              <a
                href="https://github.com/stscoundrel/old-danish-dictionary"
                rel="noopener noreferrer"
                target="_blank"
              >
                Data source
              </a>
            </li>
            <li>
              <a
                href="https://github.com/stscoundrel/old-danish-dictionary-builder"
                rel="noopener noreferrer"
                target="_blank"
              >
                Data source builder
              </a>
            </li>
          </ul>
        </nav>
        <nav
          className="nav"
        >
          <h4
            className="navTitle"
          >
            Related dictionary projects
          </h4>
          <ul>
            <li>
              <a
                href="https://old-swedish-dictionary.vercel.app/"
                rel="noopener noreferrer"
                target="_blank"
              >
                K.F Söderwall's Old Swedish Dictionary
              </a>
            </li>
            <li>
              <a
                href="https://old-norwegian-dictionary.vercel.app/"
                rel="noopener noreferrer"
                target="_blank"
              >
                Dictionary of the Old Norwegian Language
              </a>
            </li>
            <li>
              <a
                href="https://cleasby-vigfusson-dictionary.vercel.app/"
                rel="noopener noreferrer"
                target="_blank"
              >
                Cleasby and Vigfusson Old Norse Dictionary
              </a>
            </li>
            <li>
              <a
                href="https://old-icelandic.vercel.app/"
                rel="noopener noreferrer"
                target="_blank"
              >
                A Concise Dictionary of Old Icelandic
              </a>
            </li>
          </ul>
        </nav>
        <nav
          className="nav"
        >
          <h4
            className="navTitle"
          >
            Quick links
          </h4>
          <ul
            className="navColumns"
          >
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/a"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                a
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/b"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                b
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/d"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                d
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/e"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                e
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/f"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                f
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/g"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                g
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/h"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                h
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/i"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                i
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/j"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                j
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/k"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                k
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/l"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                l
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/m"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                m
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/n"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                n
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/o"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                o
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/p"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                p
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/r"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                r
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/s"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                s
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/t"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                t
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/u"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                u
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/v"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                v
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/x"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                x
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/y"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                y
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/ae"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                æ
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/oe"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                ø
              </a>
            </li>
          </ul>
        </nav>
      </div>
      <small
        className="copyright"
      >
        Copyright © 2024 - 2024
        <br />
        <a
          href="https://github.com/stscoundrel"
          rel="noopener noreferrer"
          target="_blank"
        >
          Sampo Silvennoinen / StScoundrel
        </a>
      </small>
    </div>
  </footer>,
]
`;
